src/main/java/org/skife/jdbi/v2/ClasspathStatementLocator.java

```diff
@@ -22,17 +22,60 @@
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.reflect.Method;
 import java.util.Collections;
 import java.util.Map;
 import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.regex.Matcher;
 
 /**
  * looks for [name], then [name].sql on the classpath
  */
 public class ClasspathStatementLocator implements StatementLocator
 {
-    private final Map<String, String> found = Collections.synchronizedMap(new WeakHashMap<String, String>());
+    private static class CacheKey {
+        final String name;
+        final Class<?> sqlObjectType;
+        final Method sqlObjectMethod;
+
+        CacheKey(String name, Class<?> sqlObjectType, Method sqlObjectMethod) {
+            this.name = name;
+            this.sqlObjectType = sqlObjectType;
+            this.sqlObjectMethod = sqlObjectMethod;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            CacheKey cacheKey = (CacheKey) o;
+
+            if (name != null ? !name.equals(cacheKey.name) : cacheKey.name != null) {
```

Comment:
This would be easier to read if the ternary condition were reversed:

    if (!(name == null ? cacheKey.name == null : name.equals(cacheKey.name))) {

Alternatively, add a null-safe equality method and call that in `equals`:

    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      CacheKey that = (CacheKey) o;
      return eq(this.name, that.name)
          && eq(this.sqlObjectType, that.sqlObjectType)
          && eq(this.sqlObjectMethod, that.sqlObjectMethod);
    }

    private boolean eq(Object left, Object right) {
      return left == null ? right == null : left.equals(right);
    }